CefSharp/ModelBinding/TypeSafeBinder.cs

```diff
@@ -0,0 +1,202 @@
+
+namespace CefSharp.ModelBinding
+{
+
+    /// <summary>
+    /// This class is responsible for marshaling Javascript objects into their corresponding .NET type. 
+    /// </summary>
+    /// <remarks>
+    /// This binder has no backwards compatibility with the <see cref="DefaultBinder"/> due to changes in how data member are marshaled.
+    /// </remarks>
+    public class TypeSafeBinder : IBinder
+    {
+        /// <summary>
+        /// Used to try and convert a generic type to an array via Reflection.
+        /// </summary>
+        private static readonly MethodInfo ToArrayMethodInfo = typeof(Enumerable).GetMethod("ToArray", BindingFlags.Public | BindingFlags.Static);
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="TypeSafeBinder"/> and registers converters with <see cref="TypeDescriptor"/> <br/>
+        /// This binder can reliably marshal data between the Javascript and .NET domains. <br/>
+        /// For example it provides interoperability for Typescript, Javascript, and C# style conventions without making you drop conventions in one of your languages. <br/>
+        /// It will also handle types like <see cref="Guid"/> or <see cref="Enum"/> fields that have flags. <br/>
+        /// Finally it ensure there is type safety by throwing <see cref="TypeBindingException"/> whenever data is malformed so you can catch issues in your code.
+        /// </summary>
+        public TypeSafeBinder()
+        {
+            BinderGuidConverter.Register();
+            BinderVersionConverter.Register();
+        }
+
+        /// <summary>
+        /// Attempts to bind a Javascript object into a corresponding .NET type.
+        /// </summary>
+        /// <param name="javaScriptObject">An instance of the Javascript object.</param>
+        /// <param name="nativeType">The type this method will try to create an instance of using the Javascript object.</param>
+        /// <returns>An instance of the native type.</returns>
+        public object Bind(object javaScriptObject, Type nativeType)
+        {
+            // if the intended destination is an enumeration, we can try and get the corresponding member upfront.
+            // internally this will throw a TypeBindingException if it runs into issues. See the documentation for more information.
+            if (nativeType.IsEnum)
+            {
+                return nativeType.CreateEnumMember(javaScriptObject);
+            }
+
+            // if the source object is null and is not an enum, then there isn't anything left to do.
+            if (javaScriptObject == null)
+            {
+                return null;
+            }
+            // get the underlying type for the incoming object 
+            var javaScriptType = javaScriptObject.GetType();
+
+            // if the object can be directly assigned to the destination type, then return and let the runtime handle the rest.
+            if (nativeType.IsAssignableFrom(javaScriptType))
```

Comment:
Definitely, I based it around `DefaultBinder` as not to introduce a few dozen files into this pull. I also left out generic type resoling (e.g. Model<T>) because it would have been a lot for one pull.